prov/shm/src/smr_progress.c

```diff
@@ -656,29 +635,67 @@ int smr_progress_unexp(struct smr_ep *ep,
 		entry->err = FI_EINVAL;
 	}
 
-	ret = smr_complete_rx(ep, entry->context, unexp_msg->cmd.msg.hdr.op,
-			unexp_msg->cmd.msg.hdr.op_flags | entry->flags,
-			total_len, entry->iov[0].iov_base,
-			unexp_msg->cmd.msg.hdr.addr, entry->tag,
+	comp_buf = entry->iov[0].iov_base;
+	comp_flags = (unexp_msg->cmd.msg.hdr.op_flags | entry->flags)
+			& ~SMR_MULTI_RECV;
+
+	if (entry->flags & SMR_MULTI_RECV) {
+		free_entry = smr_progress_multi_recv(ep, entry, total_len);
+		if (free_entry)
+			comp_flags |= SMR_MULTI_RECV;
+	}
+
+	ret = smr_complete_rx(ep, entry->context,
+			unexp_msg->cmd.msg.hdr.op,
+			comp_flags, total_len, comp_buf,
+			unexp_msg->cmd.msg.hdr.addr, unexp_msg->cmd.msg.hdr.tag,
 			unexp_msg->cmd.msg.hdr.data, entry->err);
 	if (ret) {
 		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 			"unable to process rx completion\n");
+		return ret;
 	}
-
 	ep->region->cmd_cnt++;
-	freestack_push(ep->unexp_fs, unexp_msg);
 
-	if (entry->flags & SMR_MULTI_RECV) {
-		ret = smr_progress_multi_recv(ep, &ep->trecv_queue, entry,
-					      total_len);
-		ret = ret ? ret : -FI_ENOMSG;
-		goto out;
+	if (free_entry) {
+		dlist_remove(&entry->entry);
+		freestack_push(ep->recv_fs, entry);
+	}
```

Comment:
This function is very similar to the one above it (from line 638 down, compared to line 383).  Try to merge?